File: draft_design_pattern/facade.nim
type
  Computer = ref object
  ComputerFacade = ref object
    computer: Computer


proc getElectricShock(self: Computer) =
  echo "ビリビリ！"

proc makeSound(self: Computer) =
  echo "ピッ！ポッ！"

proc showLoadingScreen(self: Computer) =
  echo "読み込み中..."

proc bam(self: Computer) =
  echo "準備ができました！"

proc closeEverything(self: Computer) =
  echo "ビーッ！ビーッ！ビビビビビ！"

proc sooth(self: Computer) =
  echo "（シーン）"

proc pullCurrent(self: Computer) =
  echo "プシューッ!"


proc turnOn(self: ComputerFacade) =
  self.computer.getElectricShock()
  self.computer.makeSound()
  self.computer.showLoadingScreen()
  self.computer.bam()

proc turnOff(self: ComputerFacade) =
  self.computer.closeEverything()
  self.computer.pullCurrent()
  self.computer.sooth()


proc main() =
  let computer = ComputerFacade(computer: Computer())
  computer.turnOn()
  computer.turnOff()


main()

File: draft_design_pattern/facade.ts
export {};

class Computer {
  getElectricShock(): void {
    console.log("ビリビリ！");
  }
  makeSound(): void {
    console.log("ピッ！ポッ！");
  }
  showLoadingScreen(): void {
    console.log("読み込み中...");
  }
  bam(): void {
    console.log("準備ができました！");
  }
  closeEverything(): void {
    console.log("ビーッ！ビーッ！ビビビビビ！");
  }
  sooth(): void {
    console.log("（シーン）");
  }
  pullCurrent(): void {
    console.log("プシューッ!");
  }
}

class ComputerFacade {
  constructor(private computer: Computer) {}
  turnOn(): void {
    this.computer.getElectricShock();
    this.computer.makeSound();
    this.computer.showLoadingScreen();
    this.computer.bam();
  }

  turnOff(): void {
    this.computer.closeEverything();
    this.computer.pullCurrent();
    this.computer.sooth();
  }
}

const computer = new ComputerFacade(new Computer());
computer.turnOn();
computer.turnOff();
